lib/Auth/Source/Ldap.php

```diff
@@ -49,11 +67,182 @@ public function __construct(array $info, array $config)
+            foreach ($searchAttributes as $attr) {
+                $filter .= '(' . $attr . '=' . $username . ')';
+            }
+            $filter = '(|' . $filter . ')';
+
+            // Append LDAP filters if defined
+            if ($searchFilter !== null) {
+                $filter = "(&" . $filter . "" . $searchFilter . ")";
+            }
+
+            $entry = null;
+            foreach ($searchBase as $base) {
+                $query = $ldap->query($base, $filter, $options);
+                $result = $query->execute();
+                $result = is_array($result) ? $result : $result->toArray();
+
+                if (count($result) > 1) {
+                    throw new Error\Exception(
+                        sprintf(
+                            "Library - LDAP search(): Found %d entries searching base '%s' for '%s'",
+                            count($result),
+                            $base,
+                            $filter,
+                        )
+                    );
+                } elseif (count($result) === 1) {
+                    $entry = array_pop($result);
+                    break;
+                } else {
+                    Logger::debug(
+                        sprintf(
+                            "Library - LDAP search(): Found no entries searching base '%s' for '%s'",
+                            count($result),
+                            $base,
+                            $filter,
+                        )
+                    );
+                }
+            }
+
+            if ($entry === null) {
+                throw new Error\UserNotFound("User not found");
+            }
+
+            $dn = $entry->getDn();
+        }
+
+        $ldap = $this->bind($ldapServers, $dn, $password);
+
+        $entry = null;
+        foreach ($searchBase as $base) {
+            $query = $ldap->query($base, sprintf('(distinguishedName=%s)', $dn), $options);
+            $result = $query->execute();
+            $result = is_array($result) ? $result : $result->toArray();
+
+            if (count($result) > 1) {
+                throw new Error\Exception(
+                    sprintf(
+                        "Library - LDAP search(): Found %d entries searching base '%s' for '%s'",
+                        count($result),
+                        $base,
+                        $filter,
+                    )
+                );
+            } elseif (count($result) === 1) {
+                $entry = array_pop($result);
+                break;
+            } else {
+                Logger::debug(
+                    sprintf(
+                        "Library - LDAP search(): Found no entries searching base '%s' for '%s'",
+                        count($result),
+                        $base,
+                        $filter,
+                    )
+                );
+            }
+        }
+
+        if ($entry === null) {
+            throw new Error\UserNotFound("User not found");
```

Comment:
Would be nice to make distinct strings for the two occurences of this exception so they're more easily traced to the corresponding place in the source.